server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/RawDataAggregationOperator.java

```diff
@@ -65,20 +81,78 @@ protected boolean calculateNextAggregationResult() {
         // if child still has next but can't be invoked now
         return false;
       } else {
+        if (!windowManager.isCurWindowInit()) {
+          initWindowManagerAndAggregators();
+        }
+        updateResultTsBlock();
         break;
       }
     }
 
-    // update result using aggregators
-    updateResultTsBlock();
-
     return true;
   }
 
-  private boolean calcFromRawData() {
-    Pair<Boolean, TsBlock> calcResult =
-        calculateAggregationFromRawData(inputTsBlock, aggregators, curTimeRange, ascending);
-    inputTsBlock = calcResult.getRight();
-    return calcResult.getLeft();
+  private boolean calculateAndUpdateFromRawData() {
+    if (inputTsBlock == null || inputTsBlock.isEmpty()) {
+      return false;
+    }
+
+    // if window is not initialized, we should init window status and reset aggregators
+    if (!windowManager.isCurWindowInit()) {
+      initWindowManagerAndAggregators();
+    }
+
+    if (windowManager.satisfiedCurWindow(inputTsBlock)) {
+      inputTsBlock = windowManager.skipPointsOutOfCurWindow(inputTsBlock);
+
+      int lastReadRowIndex = 0;
+      for (Aggregator aggregator : aggregators) {
+        // current agg method has been calculated
+        if (aggregator.hasFinalResult()) {
+          continue;
+        }
+
+        lastReadRowIndex = Math.max(lastReadRowIndex, aggregator.processTsBlock(inputTsBlock));
+      }
+      if (lastReadRowIndex >= inputTsBlock.getPositionCount()) {
+        inputTsBlock = null;
+        // for the last index of TsBlock, if we can know the aggregation calculation is over
+        // we can directly updateResultTsBlock and return true
+        if (isAllAggregatorsHasFinalResult(aggregators)) {
+          updateResultTsBlock();
+          return true;
+        }
+        return false;
+      } else {
+        inputTsBlock = inputTsBlock.subTsBlock(lastReadRowIndex);
+        updateResultTsBlock();
+        return true;
+      }
+    }
+    boolean isTsBlockOutOfBound = windowManager.isTsBlockOutOfBound(inputTsBlock);
+    boolean canMakeWindow = isAllAggregatorsHasFinalResult(aggregators) || isTsBlockOutOfBound;
+    if (canMakeWindow) {
+      updateResultTsBlock();
+    }
+    return canMakeWindow;
+  }
+
+  @Override
+  protected void updateResultTsBlock() {
+    appendAggregationResult(resultTsBlockBuilder, aggregators, windowManager.currentOutputTime());
+    if (windowManager.hasNext()) {
+      // reset window init status to false
+      windowManager.genNextWindow();
+    } else {
+      finish = true;
+    }
+  }
+
+  private void initWindowManagerAndAggregators() {
+    windowManager.initCurWindow(inputTsBlock);
+    IWindow curWindow = windowManager.getCurWindow();
+    for (Aggregator aggregator : aggregators) {
+      aggregator.updateWindow(curWindow);
+    }
```

Comment:
I think we only need to init IWindow once and then call some method like `moveToNextWindow` to update its window.